### PR TITLE
[BCI-2953] Initial Aptos keystore

### PR DESCRIPTION
**Pull request by yongkangchia**
## Ticket
[BCI-2953](https://smartcontract-it.atlassian.net/browse/BCI-2953)

[BCI-2953]: https://smartcontract-it.atlassian.net/browse/BCI-2953?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Changes:
* Added Aptos Keystore to Core. This includes Aptos Key which uses ED25519, Keystore, Relevant tests

Note:
* Config Related changes are pushed to Another [ticket](https://smartcontract-it.atlassian.net/browse/BCI-3511) as Aptos is still in development stages. Config changes can be made when ready.

### CODE REVIEW

**Comment by archseer** [comment]
Don't forget to include https://github.com/smartcontractkit/chainlink/blob/b7f4dc302823a7e2d9ae7f1e836c209d8294bc14/core/services/job/orm.go#L603-L606

File: core/services/chainlink/config.go
Comment (archseer) [inline core/services/chainlink/config.go]:
```
type RawConfig map[string]any
```
then use
```
Aptos []*RawConfig `toml:",omitempty"`
```
since there can be multiple configs defined

File: core/services/chainlink/config_general.go
Comment (archseer) [inline core/services/chainlink/config_general.go]: No, we want to be able to test it through LOOP.

```suggestion
func (g *generalConfig) AptosEnabled() bool {
	for _, c := range g.c.Aptos {
		v, ok := c["Enabled"]
		if !ok {
			return false
		}
		return v == true
	}
	return false
}
```

Something like this, but needs testing since `v` could be typed as `*bool`

File: core/services/chainlink/config.go
Comment (yongkangc) [reply core/services/chainlink/config.go]: thanks for the note, added this

File: core/services/chainlink/config_general.go
Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: Resolved this, thanks for the idea @archseer

**Comment by yongkangchia** [comment]
> Don't forget to include https://github.com/smartcontractkit/chainlink/blob/b7f4dc302823a7e2d9ae7f1e836c209d8294bc14/core/services/job/orm.go#L603-L606

thanks for pointing this out. there were some dependencies issues with chainwriter locally so I was planning to either add that in another pr or when all the tests pass.

File: core/services/chainlink/config.go
Comment (cfal) [inline core/services/chainlink/config.go]: aren't maps already pointers? do we need the *?

Comment (yongkangchia) [reply core/services/chainlink/config.go]: good note, thanks @cfal. maps are reference types and we don't need pointers to it

File: core/scripts/go.mod
Comment (yongkangchia) [inline core/scripts/go.mod]: This is pointing here https://github.com/smartcontractkit/chainlink-common/pull/602 for now

Will merge it in once this PR gets approved

File: core/services/keystore/keys/ocr2key/key_bundle.go
Comment (archseer) [inline core/services/keystore/keys/ocr2key/key_bundle.go:78]: Is this `new<>Keyring` method still necessary? Starknet seems to have moved to use `NewOCR2Key` directly

\cc @jmank88 (you've previously moved the starknet code into core)

File: core/services/chainlink/config.go
Comment (jmank88) [inline core/services/chainlink/config.go]: Let's dial this back - we don't actually want any chains here, it will just take some time to remove the existing:
```suggestion
// RawConfig is the config used for chains that are not embedded.
```

Comment (yongkangchia) [reply core/services/chainlink/config.go]: thanks for the note

File: core/services/keystore/keys/aptoskey/key.go
Comment (cfal) [inline core/services/keystore/keys/aptoskey/key.go]: where is this used?

File: core/services/keystore/keys/aptoskey/key.go
Comment (cfal) [inline core/services/keystore/keys/aptoskey/key.go]: the PublicKeyStr isn't currently the address

File: core/services/keystore/aptos.go
Comment (krehermann) [inline core/services/keystore/aptos.go:1]: @george-dorin what is the latest re: keystore as a loopp?

is the pattern here still the correct thing to do?

File: core/services/keystore/keys/aptoskey/key.go
Comment (yongkangc) [reply core/services/keystore/keys/aptoskey/key.go]: thanks for the note, previously used it for ID but I didn't remove it

File: core/services/keystore/keys/aptoskey/key.go
Comment (yongkangc) [reply core/services/keystore/keys/aptoskey/key.go]: thanks for this, changing the address to public key. Was using Address initially instead of public key, but right now public key makes more sense as an id if we dont want to persist and encode address.

File: core/services/keystore/aptos.go
Comment (yongkangchia) [reply core/services/keystore/aptos.go:1]: This pattern is also used for starknet which is our latest BIX integration going to mainnet soon 2 months ago

https://github.com/smartcontractkit/chainlink/pull/12575

@jmank88 mentioned something about how keystore will be extracted as a loopp in the future, maybe he can chime in here as well :)

File: core/services/keystore/keys/aptoskey/key.go
Comment (yongkangchia) [reply core/services/keystore/keys/aptoskey/key.go]: @cfal do you think we can use pbkey instead of addresses here?

Comment (yongkangchia) [reply core/services/keystore/keys/aptoskey/key.go]: Had a chat with @cfal, public key works for now. 

Modifications have to be made for aptos TXM:
https://smartcontract-it.atlassian.net/browse/BCI-3559

File: core/services/keystore/keys/ocr2key/key_bundle.go
Comment (jmank88) [reply core/services/keystore/keys/ocr2key/key_bundle.go:78]: no-memory
I don't know off the top of my head but it seems worth exploring.

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (jmank88) [inline core/services/keystore/keys/ocr2key/aptos_keyring.go]: nit/ could group libocr in its own block below these third party packages.

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (jmank88) [inline core/services/keystore/keys/ocr2key/aptos_keyring.go:9]: We try to prefer std lib `errors` when possible. Are the files you based this on still using pkg/errors?

File: core/services/keystore/keys/aptoskey/account_test.go
Comment (jmank88) [inline core/services/keystore/keys/aptoskey/account_test.go:10]: IIRC floating comments like this are dropped by the doc tools because they don't have anything to attach them to. How about making this a package comment, up at the very top?

Comment (jmank88) [reply core/services/keystore/keys/aptoskey/account_test.go:10]: Actually, that doesn't make sense since this is a test file. Maybe this is fine as-is :thinking:

File: core/services/keystore/keys/aptoskey/account.go
Comment (jmank88) [inline core/services/keystore/keys/aptoskey/account.go]: Let's move this floating comment to the end of the `AccountAddress` type.

File: core/services/keystore/aptos.go
Comment (jmank88) [reply core/services/keystore/aptos.go:1]: AFAIK following local style is still the correct thing to do.

File: core/services/keystore/keys/aptoskey/account_test.go
Comment (yongkangc) [reply core/services/keystore/keys/aptoskey/account_test.go:10]: Makes sense to have it as a package comment in my opinion, looks better

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (yongkangc) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go:9]: currently OCR2Keys are using github.com/pkg/errors, so the intention was to follow the same standards.

Should we be using std lib errors instead? if so does the other OCR2Keyrings be changed as well? If so can create another PR for that to reduce changes for this PR

Comment (yongkangc) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go:9]: currently OCR2Keys are using github.com/pkg/errors, so the intention was to follow the same standards.

Should we be using std lib errors instead? if so does the other OCR2Keyrings be changed as well? If so can create another PR for that to reduce changes for this PR

**Review by yongkangc — COMMENTED** [review]
currently OCR2Keys are using github.com/pkg/errors, so the intention was to follow the same standards.

Should we be using std lib errors instead? if so does the other OCR2Keyrings be changed as well? If so can create another PR for that to reduce changes for this PR

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (yongkangchia) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go:9]: currently OCR2Keys are using github.com/pkg/errors, so the intention was to follow the same standards.

Should we be using std lib errors instead? if so does the other OCR2Keyrings be changed as well? 

I can create another ticket/PR for that to reduce changes for this PR

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (yongkangchia) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go]: thanks for the nit!

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (jmank88) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go:9]: It is fine to match the local style for now. There are some places where we haven't given up pkg/errors because we don't want to give up the stack traces yet. We might have a different solution in the future.

**Review by jmank88 — DISMISSED** [review]

File: core/services/chainlink/config_general.go
Comment (aalu1418) [inline core/services/chainlink/config_general.go]: do we need check if the typecast doesn't fail here?

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (aalu1418) [inline core/services/keystore/keys/ocr2key/aptos_keyring.go]: might be good to check if `ok` for casting?

**Review by aalu1418 — COMMENTED** [review]
overall looks good, just a few small concerns on casting

File: core/services/keystore/keys/ocr2key/aptos_keyring.go
Comment (yongkangchia) [reply core/services/keystore/keys/ocr2key/aptos_keyring.go]: good note @aalu1418 👍🏻

File: core/services/chainlink/config_general.go
Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: for loop plugins, the type would be bool pointer. 

```
type TOMLConfig struct {
	ChainID   *string
	FeederURL *config.URL
	// Do not access directly. Use [IsEnabled]
	Enabled *bool
	Chain
	Nodes Nodes
}

func (c *TOMLConfig) IsEnabled() bool {
	return c.Enabled == nil || *c.Enabled
}
```

I dont think we need to do a type cast check? we would want the test to fail if the type ever changes

Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: maybe there is a more elegant way to handle it? since the interface for the function does not allow error to be thrown 🤔

Comment (jmank88) [reply core/services/chainlink/config_general.go]: Why don't we expand the validation to check that this field can be cast to bool?

**Review by cfal — DISMISSED** [review]

File: core/services/chainlink/config_general.go
Comment (aalu1418) [reply core/services/chainlink/config_general.go]: @jmank88 could be good!
it's just one of the things i look for when I see casting to make sure something can't panic if the user does something weird

Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: thanks for the feedback @aalu1418 & @jmank88 added typecasting for *bool and bool

and using panic to make panic explicit instead of implicit

**Comment by yongkangchia** [comment]
Around Config validation of AptosEnabled: added typecasting using switch for *bool and bool, using panic to make panic explicit instead of implicit

File: core/services/chainlink/config_general.go
Comment (jmank88) [inline core/services/chainlink/config_general.go]: Sorry I should have said more. If you add a `ValidateConfig` method to `RawConfig`, then you can return a user friendly error: https://github.com/smartcontractkit/chainlink-common/blob/main/pkg/config/validate.go#L17

Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: thanks for this reference @jmank88 

Do you mean something like this?

```
// Validate returns an error if the Config is not valid for use, as-is.
func (c *RawConfig) ValidateConfig() error {
	if v, ok := (*c)["Enabled"]; ok {
		switch v.(type) {
		case nil, *bool:
			return nil
		default:
			return fmt.Errorf("invalid type for 'Enabled': expected nil, bool, or *bool, got %T", v)
		}
	}

	return nil
}
```

And this `ValidateConfig()` would be called by the validate function of `generalConfig` [here](https://github.com/smartcontractkit/chainlink/blob/eace378772fe88db6c33f93b36d1ad2a421b339e/core/services/chainlink/config_general.go#L219)?

and so the logic for the AptosEnabled check does not need to account for typecast checking

**Review by archseer — DISMISSED** [review]

File: core/services/chainlink/config.go
Comment (archseer) [inline core/services/chainlink/config.go]: This should use the ` RawConfigs` type from below but I'm not sure if the `RawConfig` type should be as a pointer or not

Comment (cfal) [reply core/services/chainlink/config.go]: +1, this was updated previously https://github.com/smartcontractkit/chainlink/pull/13564#discussion_r1646585773 - was there a reason we needed it to be pointers again?

File: core/services/chainlink/config.go
Comment (jmank88) [inline core/services/chainlink/config.go]: You can test this by adding an invalid value here: https://github.com/smartcontractkit/chainlink/blob/a830fe093f4642236a89ac73bf9eefe5008898bc/core/services/chainlink/config_test.go#L1251-L1252

Notice the error messages are formatted. You should return one of these well known errors that include the field name: https://github.com/smartcontractkit/chainlink/blob/a830fe093f4642236a89ac73bf9eefe5008898bc/core/chains/evm/config/toml/config.go#L923-L933

File: core/services/chainlink/config.go
Comment (yongkangchia) [reply core/services/chainlink/config.go]: @archseer thanks for the note, RawConfig type does not need to be a pointer. I made mistakes in merging previously 🙏🏻 @cfal

File: core/services/chainlink/config.go
Comment (yongkangchia) [reply core/services/chainlink/config.go]: thanks for this suggestion, this is much more elegant way. noted on testing as well

**Comment by yongkangchia** [comment]
Added tests to validation of configs. However the invalidation tests are currently formatted slightly differently because its just one error.

```
Now:     - Aptos.0.Enabled: invalid value (1): expected *bool

With more tests:
                                +       - Aptos.0: 2 errors:
                                +                       - Enabled: invalid value (1): expected *bool
                                +                       - test: invalid value (1): this is a test
```

File: core/services/chainlink/config_test.go
Comment (yongkangchia) [inline core/services/chainlink/config_test.go:1361]: This is formatted this way due to one test

```
Now:     - Aptos.0.Enabled: invalid value (1): expected *bool

With more tests:
                                +       - Aptos.0: 2 errors:
                                +                       - Enabled: invalid value (1): expected *bool
                                +                       - test: invalid value (1): this is a test
```

File: core/web/auth/auth_test.go
Comment (yongkangchia) [inline core/web/auth/auth_test.go:291]: starknet is missing, we need to add this into starknet

File: core/services/chainlink/config_general.go
Comment (jmank88) [inline core/services/chainlink/config_general.go]: What is left to update?

File: core/services/chainlink/testdata/config-invalid.toml
Comment (jmank88) [inline core/services/chainlink/testdata/config-invalid.toml]: What is this extra one for?

File: core/services/keystore/keys/aptoskey/account.go
Comment (jmank88) [inline core/services/keystore/keys/aptoskey/account.go]: Let's move this up to `type AccountAddress` as a second paragraph on the comment. Otherwise the tooling doesn't know where to put it when generating the docs.

File: core/services/chainlink/testdata/config-invalid.toml
Comment (yongkangchia) [reply core/services/chainlink/testdata/config-invalid.toml]: Looking at the other Non-evm config, it seems using empty array tables to test whether the config becomes invalidated without a config. Thought it might be testing the edge case that's why i added it in 

i.e Solana
https://github.com/smartcontractkit/chainlink/blob/3f8c00a6f1884f765bbe9e4b70e0dc4fb94a0088/core/services/chainlink/testdata/config-invalid.toml#L133-L147

File: core/services/keystore/keys/aptoskey/account.go
Comment (yongkangchia) [reply core/services/keystore/keys/aptoskey/account.go]: thanks for the note @jmank88, added

File: core/services/chainlink/config_general.go
Comment (yongkangchia) [reply core/services/chainlink/config_general.go]: thanks for pointing this out @jmank88 . Removing this from the code

File: core/services/chainlink/testdata/config-invalid.toml
Comment (jmank88) [reply core/services/chainlink/testdata/config-invalid.toml]: Is there a corresponding failing test case? If not, then it can be removed.

File: core/services/keystore/keys/aptoskey/account.go
Comment (jmank88) [inline core/services/keystore/keys/aptoskey/account.go]:
```suggestion
// It can represent an Object, an Account, and much more.
//
// AccountAddress is copied from the aptos sdk because:
//  1. There are still breaking changes in sdk and we don't want the dependency.
//  2. AccountAddress is just a wrapper and can be easily extracted out.
// https://github.com/aptos-labs/aptos-go-sdk/blob/main/internal/types/account.go
```

File: core/services/chainlink/testdata/config-invalid.toml
Comment (yongkangchia) [reply core/services/chainlink/testdata/config-invalid.toml]: There isn't a corresponding failing test case for now, removed it. thanks for the suggestion

**Review by jmank88 — APPROVED** [review]

**Review by archseer — APPROVED** [review]

**Pull request merged**
